components/builder-graph/src/data_store.rs

```diff
@@ -25,16 +32,181 @@ use crate::{config::Config,
+                     RunQueryDsl},
             error::{Error,
-                    Result}};
+                    Result},
+            package_info::PackageInfo};
+
+use crate::package_ident_intern::PackageIdentIntern;
+
+// Take a list of packages and return the set that is unbuildable
+// Separate from the below DataStoreTrait to minimize leakage into the solver
+pub trait Unbuildable: AsUnbuildable {
+    fn filter_unbuildable(&self,
+                          packages: &[PackageIdentIntern],
+                          _target: PackageTarget)
+                          -> Result<Vec<PackageIdentIntern>> {
+        Ok(filter_unbuildable_static(&packages))
+    }
+}
+
+// These two exist because trait upcasting isn't a thing in rust, so we have to do some magic
+// instead. Thank you stack overflow https://stackoverflow.com/questions/28632968/why-doesnt-rust-support-trait-object-upcasting
+pub trait AsUnbuildable {
+    fn as_unbuildable(&self) -> &dyn Unbuildable;
+}
+
+impl<T: Unbuildable> AsUnbuildable for T {
+    fn as_unbuildable(&self) -> &dyn Unbuildable { self }
+}
+
+// Capture the key database APIs and separate them from implementation to make testing with static
+// data easier.
+pub trait DataStoreTrait: Unbuildable {
+    fn get_job_graph_packages(&self) -> Result<Vec<PackageInfo>>;
+
+    fn get_job_graph_package(&self,
+                             ident: &PackageIdent,
+                             target: PackageTarget)
+                             -> Result<PackageWithVersionArray>;
+
+    fn get_origin_channel_latest(&self,
+                                 origin: &str,
+                                 channel: &str,
+                                 target: PackageTarget)
+                                 -> Result<Vec<PackageIdent>>;
+
+    fn serialize(&self,
+                 filename: &str,
+                 origin: &str,
+                 channel: &str,
+                 target: PackageTarget)
+                 -> Result<()>;
+}
+
+/// Simple serial
+#[derive(Debug, Serialize, Deserialize)]
+pub struct SerializedDatabase {
+    origin:        String,
+    channel:       String,
+    target:        PackageTarget,
+    packages:      Vec<PackageWithVersionArray>,
+    base_packages: Vec<PackageIdent>,
+}
+
+impl SerializedDatabase {
+    pub fn write_to_file(&self, filename: &str) -> Result<()> {
+        let path = Path::new(filename);
+        let mut file = File::create(&path).unwrap();
+
+        let serialized = serde_json::to_string(&self).unwrap();
+        file.write_all(serialized.as_bytes()).unwrap();
+        Ok(())
+    }
+
+    pub fn read_from_file(filename: &str) -> Result<SerializedDatabase> {
+        let path = Path::new(filename);
+        let file = File::open(&path)?;
+        let reader = BufReader::new(file);
+        let u: SerializedDatabase = serde_json::from_reader(reader)?;
+        Ok(u)
+    }
+
+    pub fn package_count(&self) -> usize { self.packages.len() }
+}
+
+#[cfg(test)]
```

Comment:
Just a nit - In most of our other components where we have unit tests we've typically created an aggregate test section at the bottom of the functional code. Is there a particular style reason we have this here as opposed to the end of the file?